presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeJoinOptimizer.java

```diff
@@ -0,0 +1,142 @@
+import com.facebook.presto.spi.relation.VariableReferenceExpression;
+import com.facebook.presto.sql.parser.SqlParser;
+import com.facebook.presto.sql.planner.PlanVariableAllocator;
+import com.facebook.presto.sql.planner.TypeProvider;
+import com.facebook.presto.sql.planner.plan.ExchangeNode;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.facebook.presto.sql.planner.plan.SimplePlanRewriter;
+
+import java.util.List;
+
+import static com.facebook.presto.SystemSessionProperties.preferMergeJoin;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+public class MergeJoinOptimizer
+        implements PlanOptimizer
+{
+    private final Metadata metadata;
+    private final SqlParser parser;
+
+    public MergeJoinOptimizer(Metadata metadata, SqlParser parser)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.parser = requireNonNull(parser, "parser is null");
+    }
+
+    @Override
+    public PlanNode optimize(PlanNode plan, Session session, TypeProvider type, PlanVariableAllocator variableAllocator, PlanNodeIdAllocator idAllocator, WarningCollector warningCollector)
+    {
+        requireNonNull(plan, "plan is null");
+        requireNonNull(session, "session is null");
+        requireNonNull(variableAllocator, "variableAllocator is null");
+        requireNonNull(idAllocator, "idAllocator is null");
+
+        if (preferMergeJoin(session)) {
+            return SimplePlanRewriter.rewriteWith(new MergeJoinOptimizer.Rewriter(variableAllocator, idAllocator, metadata, session), plan, null);
+        }
+        return plan;
+    }
+
+    private class Rewriter
+            extends SimplePlanRewriter<Void>
+    {
+        private final PlanNodeIdAllocator idAllocator;
+        private final Metadata metadata;
+        private final Session session;
+        private final TypeProvider types;
+
+        private Rewriter(PlanVariableAllocator variableAllocator, PlanNodeIdAllocator idAllocator, Metadata metadata, Session session)
+        {
+            this.idAllocator = requireNonNull(idAllocator, "idAllocator is null");
+            this.metadata = requireNonNull(metadata, "metadata is null");
+            this.session = requireNonNull(session, "session is null");
+            this.types = variableAllocator.getTypes();
+        }
+
+        @Override
+        public PlanNode visitJoin(JoinNode node, RewriteContext<Void> context)
+        {
+            // When we have a plan that looks like:
+            // JoinNode
+            //- TableScanA
+            //- LocalExchange
+            //  - TableScanB
+
+            // We check the data properties of TableScanA and TableScanB to see if they meet requirements for merge join:
+            // 1. If so, we remove the extra LocalExchange and add mergeJoinEnabled true in JoinNode, plan would be
+            // JoinNode (mergeJoin enabled)
```

Comment:
Should we consider adding a new JoinNode type?  Some aspects of `JoinNode` seem specific to hash join, like `dynamicFilters`.  `DistributionType` also seems like it might not make much sense, since we wouldn't have broadcast join for merge join (right?).  For reference, there is a separate `IndexJoinNode` type for index join.

Reply:
This is a good point. Eventually, I think we do want to add a new JoinNode for merge join. But currently it is not supported on the execution side yet. Reusing the existing JoinNode allows the worker to fall back to use Hash Join operator when merge join operator is not implemented. 

I think maybe we can work on introducing a new JoinNode and remove the fall back path when both Java and CPP workers add the merge join implementations. What do you think, Tim?

Reply:
I would say we shouldn't be planning a merge join if it's not supported in the execution, and we should fail in the execution if we create a mergejoin in the planner and it's not supported in the execution yet.   Otherwise I feel we are just asking for problems.  e.g. here we're removing a local exchange when we have mergeJoinEnabled,even though we would want that local exchange if we were to fall back to a hash join.

Reply:
I think we agree on the final state to add a new JoinNode. I don't have a strong preference on the intermediate state about whether to reject the query or fall back to hash join, as we are not supposed to enable it before the execution part is ready.
Rejecting the query seems a bit cleaner. We can go with that approach.

Reply:
How about the implementation on CPP side as of now? I remember they were using the original JoinNode

To unblock the current test, as of now, we can add it in java execution side when we see mergeJoinEnabled true, we reject the query. 

Then when we move to a new JoinNode type as a final state, we make changes to plan side and execution side from both CPP and Java

@yuanzhanhku @rschlussel thoughts?

Reply:
CPP side needs changes anyway for either approach. We have a cpp side prototype. But that is not submitted yet. This planning side change needs to go in first such that the CPP side can create the merge join operator according to the plan change. I think the cpp side complexity of the two approaches are similar, i.e., create the merge join operator if the new bool mergeJoinEnabled is set to true in the original join node or if it is a new join node.

So the cpp side change is not a blocker for introducing a new join node type.